core/packages/contracts/src/NativeTokens.sol

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.19;
+
+import "openzeppelin/access/Ownable.sol";
+import "openzeppelin/access/AccessControl.sol";
+import "openzeppelin/token/ERC20/extensions/IERC20Metadata.sol";
+
+import "./TokenVault.sol";
+import "./SubstrateTypes.sol";
+import "./NativeTokensTypes.sol";
+import "./OutboundChannel.sol";
+
+/// @title Native Tokens
+/// @dev Manages locking, unlocking ERC20 tokens in the vault. Initializes ethereum native
+/// tokens on the substrate side via create.
+contract NativeTokens is AccessControl {
+    /// @dev Describes the type of message.
+    enum Action {
+        Unlock
+    }
+
+    /// @dev Message format.
+    struct Message {
+        /// @dev The action type.
+        Action action;
+        /// @dev The message payload.
+        bytes payload;
+    }
+
+    /// @dev Unlock payload format.
+    struct UnlockPayload {
+        /// @dev The ERC20 token to unlock.
+        address token;
+        /// @dev The destination address that will receive unlocked funds.
+        address recipient;
+        /// @dev The amount to unlock.
+        uint128 amount;
+    }
+
+    /// @dev Emitted once the funds are locked and a message is successfully queued.
+    event Locked(bytes recipient, address token, uint128 amount);
+
+    /// @dev Emitted once the funds are unlocked.
+    event Unlocked(address recipient, address token, uint128 amount);
+
+    /// @dev Emitted after enqueueing a a create token message to substrate.
+    event Created(address token);
+
+    /// @dev Set a new outbound channel.
+    event OutboundChannelUpdated(address newOutBoundChannel);
+
+    /* State */
+
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+    bytes32 public constant SENDER_ROLE = keccak256("SENDER_ROLE");
+
+    bytes32 public immutable peerID;
+    bytes public peer;
+
+    TokenVault public immutable vault;
+    IOutboundChannel public outboundChannel;
+
+    uint256 public createTokenFee;
+
+    /* Errors */
+
+    error InvalidAmount();
+    error Unauthorized();
+    error NoFundsforCreateToken();
+
+    constructor(
+        TokenVault _vault,
+        IOutboundChannel _outboundChannel,
+        bytes memory _peer,
+        uint256 _createTokenFee
+    ) {
+        _grantRole(ADMIN_ROLE, msg.sender);
+        _setRoleAdmin(SENDER_ROLE, ADMIN_ROLE);
+        vault = _vault;
+        outboundChannel = _outboundChannel;
```

Comment:
are we gonna also to add `inboundChannel` here with `SENDER_ROLE`? 
detail comment here.
https://github.com/Snowfork/snowbridge/pull/781/files#r1135097810

Reply:
yes, that's correct. We set that up here in `Deploy.sol`:
https://github.com/Snowfork/snowbridge/pull/781/files#diff-5b435e0b60b4b5dae9b667651a36e777d2de6372627be4a9fea0d9d7986565adR55